api/lib/domain/services/authentication/authentication-service-registry.js

```diff
@@ -1,14 +1,29 @@
-import * as OidcIdentityProviders from '../../constants/oidc-identity-providers.js';
 import { InvalidIdentityProviderError } from '../../errors.js';
+import { PoleEmploiOidcAuthenticationService } from './pole-emploi-oidc-authentication-service.js';
+import { CnavOidcAuthenticationService } from './cnav-oidc-authentication-service.js';
+import { FwbOidcAuthenticationService } from './fwb-oidc-authentication-service.js';
 
-function lookupAuthenticationService(identityProvider) {
-  const identityProviderService = Object.values(OidcIdentityProviders).find(
-    (oidcIdentityProvider) => oidcIdentityProvider.service.code === identityProvider
-  );
+const OidcProviderServiceMap = [
+  new PoleEmploiOidcAuthenticationService(),
+  new CnavOidcAuthenticationService(),
+  new FwbOidcAuthenticationService(),
+].reduce((accumulator, service) => {
+  return {
+    ...accumulator,
+    [service.code]: service,
+  };
+}, {});
```

Comment:
question : j'ai du mal à comprendre ce que retourne le `.reduce()`, ça serait possible d'avoir plus d'explication 🙏🏼

Reply:
[`.reduce()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce) est une méthode très puissante de `Array`, mais pas immédiatement facile à comprendre et à lire.

Le `reduce` dans le cas du code ci-dessus sert, à partir d'un tableau de *OidcAuthenticationServices*, à créer un `Object` (avec un mapping clé-valeur, avec les clés qui sont le `code` *OidcAuthenticationServices* et les valeurs qui sont des *OidcAuthenticationServices*), d'où le nom `OidcProviderServiceMap`. Cela permet de disposer d'un `Object`  `OidcProviderServiceMap` « puissant sémantiquement » dans lequel on pourra récupérer facilement plus tard selon les besoins les services eux-mêmes ou bien les identifiants des services (leurs `code`).

Et ceci correspond au cas « array reduce with an object as accumulator » qui est traité notamment sur SO ici :
https://stackoverflow.com/questions/53581550/array-reduce-with-an-object-as-accumulator

Reply:
Merci beaucoup ! J'ai encore deux petites questions : 

1. Mais du coup qu'est-ce qu'il y a dans `accumulator` ?
2. La méthode va retourner par exemple : `{ CNAV : new CnavOidcAuthenticationService()}` ?

Reply:
Pour le point **2.**, oui c'est bien ça 👍

Pour le point **1.**, l'accumulator est passé à chaque itération de la méthode `reduce()` et cela donne donc les itérations ci-dessous :

Au départ accumulator vaut : 
```javascript
{}
```
C'est la ligne `15` qui passe `{}` à la méthode `reduce()`.

À la 1ère itération accumulator vaut :
```javascript
{'POLE_EMPLOI': new PoleEmploiOidcAuthenticationService()}
```

À la 2ème itération accumulator vaut : 
```javascript
{'POLE_EMPLOI': new PoleEmploiOidcAuthenticationService(), 'CNAV' : new CnavOidcAuthenticationService()}
```

À la 3ème et dernière itération accumulator vaut : 
```javascript
{'POLE_EMPLOI': new PoleEmploiOidcAuthenticationService(), 'CNAV' : new CnavOidcAuthenticationService(), 'FWB': new FwbOidcAuthenticationService()}
```

Reply:
Lire du code `reduce` me fait toujours des nœuds au cerveau 🤯

Peut-être qu'on pourrait écrire ce code sur plusieurs lignes et peut-être nommer la fonction anonyme pour faciler la lecture ?

Reply:
`reduce` est la méthode/fonction que je trouve la plus compliquée en JavaScript. Mais elle est puissante et bien adaptée pour traiter certains cas :-)

Reply:
Proposition concrète pour rendre le code plus lisible :

```suggestion
const oidcProviderServices = [
  new PoleEmploiOidcAuthenticationService(),
  new CnavOidcAuthenticationService(),
  new FwbOidcAuthenticationService(),
];

const oidcProviderServicesMap = OidcProviderServices.reduce(_associateServiceToCode, {});
function _associateServiceToCode(map, service) {
  return {
    ...map,
    [service.code]: service,
  };
}
```

Reply:
Le code a été rendu normalement plus facile à comprendre en adoptant la proposition concrète ci-dessus.